File: packages/rust-sdk/src/operations/log_drains.rs
use crate::client::StraitClient;
use crate::errors::StraitError;
use crate::http::substitute_path_params;
use std::sync::Arc;

pub struct LogDrainsService {
    client: Arc<StraitClient>,
}

impl LogDrainsService {
    pub fn new(client: Arc<StraitClient>) -> Self {
        Self { client }
    }

    pub async fn list(
        &self,
        query: Option<&[(&str, &str)]>,
    ) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("GET", "/v1/log-drains", query, None, None)
            .await
    }

    pub async fn create(&self, body: serde_json::Value) -> Result<serde_json::Value, StraitError> {
        self.client
            .do_request("POST", "/v1/log-drains", None, None, Some(body))
            .await
    }

    pub async fn get(&self, drain_id: &str) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/log-drains/{drainID}", &[("drainID", drain_id)]);
        self.client.do_request("GET", &path, None, None, None).await
    }

    pub async fn update(
        &self,
        drain_id: &str,
        body: serde_json::Value,
    ) -> Result<serde_json::Value, StraitError> {
        let path = substitute_path_params("/v1/log-drains/{drainID}", &[("drainID", drain_id)]);
        self.client
            .do_request("PATCH", &path, None, None, Some(body))
            .await
    }

    pub async fn delete(&self, drain_id: &str) -> Result<(), StraitError> {
        let path = substitute_path_params("/v1/log-drains/{drainID}", &[("drainID", drain_id)]);
        self.client
            .do_request_no_content("DELETE", &path, None, None, None)
            .await
    }
}

File: packages/go-sdk/operations/log_drains.go
package operations

import "context"

// LogDrainsService provides log drain management operations.
type LogDrainsService struct{ r Requester }

func NewLogDrainsService(r Requester) *LogDrainsService { return &LogDrainsService{r: r} }

func (s *LogDrainsService) List(ctx context.Context, query map[string]string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", "/v1/log-drains", query, nil, nil, &result)
	return result, err
}

func (s *LogDrainsService) Create(ctx context.Context, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "POST", "/v1/log-drains", nil, nil, body, &result)
	return result, err
}

func (s *LogDrainsService) Get(ctx context.Context, drainID string) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "GET", PathParams("/v1/log-drains/{drainID}", map[string]string{"drainID": drainID}), nil, nil, nil, &result)
	return result, err
}

func (s *LogDrainsService) Update(ctx context.Context, drainID string, body any) (map[string]any, error) {
	var result map[string]any
	err := s.r.DoRequest(ctx, "PATCH", PathParams("/v1/log-drains/{drainID}", map[string]string{"drainID": drainID}), nil, nil, body, &result)
	return result, err
}

func (s *LogDrainsService) Delete(ctx context.Context, drainID string) error {
	return s.r.DoRequest(ctx, "DELETE", PathParams("/v1/log-drains/{drainID}", map[string]string{"drainID": drainID}), nil, nil, nil, nil)
}
